materialize-databricks/client.go

```diff
@@ -12,98 +13,111 @@ import (
-
-	rows, err := c.db.QueryContext(ctx, fmt.Sprintf(`
-		select table_schema, table_name, column_name, is_nullable, data_type, character_maximum_length, column_default
-		from system.information_schema.columns
-		where table_catalog = %s
-		and table_schema in (%s);
-	`,
-		c.ep.Dialect.Literal(c.cfg.CatalogName),
-		strings.Join(schemas, ","),
-	))
-	if err != nil {
-		return nil, fmt.Errorf("querying system.information_schema.columns: %w", err)
+	// Databricks' Tables API provides a free-form "metadata" field that is a JSON object containing
+	// some additional useful information.
+	type columnMeta struct {
+		Metadata struct {
+			Default json.RawMessage `json:"CURRENT_DEFAULT,omitempty"`
+		} `json:"metadata"`
 	}
-	defer rows.Close()
-
-	type columnRow struct {
-		TableSchema            string
-		TableName              string
-		ColumnName             string
-		IsNullable             string
-		DataType               string
-		CharacterMaximumLength stdsql.NullInt64
-		ColumnDefault          stdsql.NullString
+
+	// The table listing will fail if the schema doesn't already exist, so only attempt to list
+	// tables in schemas that do exist.
+	existingSchemas, err := c.ListSchemas(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("listing schemas: %w", err)
 	}
 
-	for rows.Next() {
-		var c columnRow
-		if err := rows.Scan(&c.TableSchema, &c.TableName, &c.ColumnName, &c.IsNullable, &c.DataType, &c.CharacterMaximumLength, &c.ColumnDefault); err != nil {
-			return nil, fmt.Errorf("scanning column row: %w", err)
+	for sc := range rpSchemas {
+		if !slices.Contains(existingSchemas, sc) {
+			log.WithField("schema", sc).Debug("not listing tables for schema since it doesn't exist")
+			continue
 		}
 
-		is.PushField(boilerplate.EndpointField{
-			Name:               c.ColumnName,
-			Nullable:           strings.EqualFold(c.IsNullable, "yes"),
-			Type:               c.DataType,
-			CharacterMaxLength: int(c.CharacterMaximumLength.Int64),
-			HasDefault:         c.ColumnDefault.Valid,
-		}, c.TableSchema, c.TableName)
-	}
+		tableIter := c.wsClient.Tables.List(ctx, catalog.ListTablesRequest{
+			CatalogName: c.cfg.CatalogName,
+			SchemaName:  sc,
+		})
+
+		for tableIter.HasNext(ctx) {
+			t, err := tableIter.Next(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("iterating tables: %w", err)
+			}
 
-	if err := rows.Err(); err != nil {
-		return nil, fmt.Errorf("rows.Err(): %w", err)
+			is.PushResource(t.SchemaName, t.Name)
+
+			for _, c := range t.Columns {
+				var colMeta columnMeta
+				if err := json.Unmarshal([]byte(c.TypeJson), &colMeta); err != nil {
+					return nil, fmt.Errorf("unmarshalling column metadata: %w", err)
+				}
+
+				is.PushField(boilerplate.EndpointField{
+					Name:               c.Name,
+					Nullable:           c.Nullable,
+					Type:               string(c.TypeName),
+					CharacterMaxLength: 0, // TODO(whb): Currently not supported by us, although we could parse the metadata for VARCHAR columns.
```

Comment:
Is this a regression from the last version given that we used to have this information? I'm not sure what does this property do / what are the implications of not having it now

Reply:
As far as I can tell databricks would always say 0 for the "CharacterMaximumLength" even from querying it out of `INFORMATION_SCHEMA.COLUMNS`. A way I could see us getting this information would be to parse it out of the "metadata" JSON returned by this API, which will tell you if a field is for example a `VARCHAR(20)`.

But we only use the `CharacterMaxLength` for `materialize-redshift` and `materialize-mysql` right now, for them to dynamically resize columns, because they don't support an efficient string column of arbitrary length. So since the `CharacterMaxLength` is not used by our Databricks materialization I didn't think it was worth the effort to implement the parsing now.